morph_tool/plot/morphology.py

```diff
@@ -0,0 +1,99 @@
+"""Module for drawing morphologies with synapses."""
+
+from functools import partial
+import warnings
+
+import pandas as pd
+from pandas import Categorical
+from pandas.core.dtypes.common import is_integer_dtype
+from plotly import express
+
+from neurom import morphmath
+from neurom.view.plotly import get_figure
+from morph_tool.plot.consts import (PRE_SECTION_ID, PRE_SEGMENT_ID, PRE_SEGMENT_OFFSET,
+                                    POST_SECTION_ID, POST_SEGMENT_ID, POST_SEGMENT_OFFSET)
+
+REQUIRED_PRE_COLUMNS = {PRE_SECTION_ID, PRE_SEGMENT_ID, PRE_SEGMENT_OFFSET}
+REQUIRED_POST_COLUMNS = {POST_SECTION_ID, POST_SEGMENT_ID, POST_SEGMENT_OFFSET}
+
+
+def _validate_synapses(synapses):
+    def _is_int(*columns):
+        return all(is_integer_dtype(synapses[c].dtype) for c in columns)
+
+    is_pre = REQUIRED_PRE_COLUMNS.issubset(synapses.columns)
+    is_post = REQUIRED_POST_COLUMNS.issubset(synapses.columns)
+    assert is_pre or is_post, \
+        f'{REQUIRED_PRE_COLUMNS} or {REQUIRED_POST_COLUMNS} are required columns of `synapses`'
+    if is_pre and not _is_int(PRE_SECTION_ID, PRE_SEGMENT_ID) or \
+            is_post and not _is_int(POST_SECTION_ID, POST_SEGMENT_ID):
+        warnings.warn('Section ids and segment ids columns of `synapses` arg are not integers, and '
+                      'will be forced to integer type')
+
+
+def _add_coords(synapse, morphology):
+    """Adds coordinates and direction fields to ``synapses`` via ``apply`` function."""
+    is_pre = PRE_SECTION_ID in synapse.index and not pd.isnull(synapse[PRE_SECTION_ID])
+    is_post = POST_SECTION_ID in synapse.index and not pd.isnull(synapse[POST_SECTION_ID])
+    assert is_pre ^ is_post, 'Synapse must have either afferent or efferent section ids for the ' \
```

Comment:
I think `is_pre != is_post` would be more clear. The `^` operator is not well know IMO.

Reply:
thank you, it is clearer with `!=`

Reply:
you're welcome :)